src/transformers/hf_argparser.py

```diff
@@ -0,0 +1,110 @@
+import dataclasses
+from argparse import ArgumentParser
+from enum import Enum
+from typing import Any, Iterable, NewType, Tuple, Union
+
+
+DataClass = NewType("DataClass", Any)
+DataClassType = NewType("DataClassType", Any)
+
+
+class HfArgumentParser(ArgumentParser):
+    """
+    This subclass of `argparse.ArgumentParser` uses type hints on dataclasses
+    to generate arguments.
+    """
+
+    def __init__(self, obj: Union[None, DataClass, DataClassType] = None):
+        """
+        Args:
+            obj:
+                (Optional) Can be either a dataclass instance, or a dataclass type.
+                Both will work the exact same way.
+        """
+        super().__init__()
+        if obj is not None:
+            self.add_dataclass_arguments(obj)
+
+    def add_dataclass_arguments(self, obj: Union[DataClass, DataClassType]):
```

Comment:
Maybe accept a list of Dataclasses as well?

Name could be also `add_arguments_from_dataclasses`

Reply:
Store the input Type in a `self.dataclasse_types` to generate dataclasses when parsing?